src/Mod/Arch/ArchRoof.py

```diff
@@ -286,393 +383,455 @@ def makeRoofProfilsDic(self, id, angle, run, idrel, overhang, thickness,):
-        edgeHip = DraftGeomUtils.edg(FreeCAD.Vector(ptInterRidges[0]),profilCurrent["edge"].Vertexes[0].Point)
-        ptInterHipEave = DraftGeomUtils.findIntersection(edgeHip,profilCurrent["eave"],infinite1=True,infinite2=False,)
-        if ptInterHipEave:
-            self.ptsPaneProject.append(FreeCAD.Vector(ptInterHipEave[0]))
-        else:
-            ptInterHipEave = DraftGeomUtils.findIntersection(edgeHip,profilBack1["eaveD"],infinite1=True,infinite2=True,)
-            self.ptsPaneProject.append(FreeCAD.Vector(ptInterHipEave[0]))
-            self.ptsPaneProject.append(FreeCAD.Vector(profilCurrent["eave"].Vertexes[0].Point))
-
-    def nextHigher(self, i):
-
-        import DraftGeomUtils
-        profilCurrent = self.findProfil(i)
-        profilNext1 = self.findProfil(i+1)
-        dec = profilCurrent["height"]/math.tan(math.radians(profilNext1["angle"]))
-        edgeRidgeOnPane = DraftGeomUtils.offset(profilNext1["edge"],self.getPerpendicular(profilNext1["vec"],profilNext1["rot"],dec))
-        ptInter = DraftGeomUtils.findIntersection(profilCurrent["ridge"],edgeRidgeOnPane,infinite1=True,infinite2=True,)
-        edgeHip = DraftGeomUtils.edg(FreeCAD.Vector(ptInter[0]),profilCurrent["edge"].Vertexes[-1].Point)
-        ptInterHipEave = DraftGeomUtils.findIntersection(edgeHip,profilCurrent["eave"],infinite1=True,infinite2=False,)
-        if ptInterHipEave:
-            self.ptsPaneProject.append(FreeCAD.Vector(ptInterHipEave[0]))
-        else:
-            ptInterHipEave = DraftGeomUtils.findIntersection(edgeHip,profilNext1["eaveD"],infinite1=True,infinite2=True,)
-            self.ptsPaneProject.append(FreeCAD.Vector(profilCurrent["eave"].Vertexes[-1].Point[0]))
-            self.ptsPaneProject.append(FreeCAD.Vector(ptInterHipEave[0]))
-        self.ptsPaneProject.append(FreeCAD.Vector(ptInter[0]))
-        ptInterRidges = DraftGeomUtils.findIntersection(profilCurrent["ridge"],profilNext1["ridge"],infinite1=True,infinite2=True,)
-        self.ptsPaneProject.append(FreeCAD.Vector(ptInterRidges[0]))
+    def nextSameHeight(self, i):
+        profilCurr = self.findProfil(i)
+        profilNext = self.findProfil(i + 1)
+        self.helperSloped(profilCurr,
+                          profilNext,
+                          profilCurr["ridge"],
+                          profilNext["ridge"],
+                          isBack = False)
 
     def backHigher(self, i):
+        profilCurr = self.findProfil(i)
+        profilBack = self.findProfil(i - 1)
+        dec = profilCurr["height"] / math.tan(math.radians(profilBack["angle"]))
+        per = self.getPerpendicular(profilBack["vec"], profilBack["rot"], dec)
+        edgeRidgeOnPane = DraftGeomUtils.offset(profilBack["edge"], per)
+        self.helperSloped(profilCurr,
+                          profilBack,
+                          profilCurr["ridge"],
+                          edgeRidgeOnPane,
+                          isBack = True)
```

Comment:
Okay, so here, the fourth element `edgeRidgeOnPane` is not `profilBack["ridge"]` so this is a special case. However, it is still derived from `profilBack`, so maybe there is a different way of handling this.

Reply:
Aha, you now understand the algo a little bit. Sure there are probably dozens of ways to do this. This is just my attempt to make the code more efficient. Instead of 6 functions that contained more or less the same code they now use a single helper function, making the code much easier to maintain.

Reply:
Indeed. Maybe something here can still be refactored at some other point.